File: Story 3/Task5.cpp
/**
 * @filename - Task-5.cpp
 * @description - The Prime/Composite List Partition 
 * @author - Disha Toshniwal 
 */
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

struct Node {
    int value;
    Node* next;
    Node(int v) : value(v), next(nullptr) {}
};

// Check if a number is prime
bool isPrime(int number) {
    if (number <= 1) return false;
    if (number == 2) return true;
    if (number % 2 == 0) return false;

    for (int divisor = 3; divisor * divisor <= number; divisor += 2)
        if (number % divisor == 0) return false;

    return true;
}

Node* partitionPrimes(Node* head) {

    Node* primeHead = nullptr;
    Node* primeTail = nullptr;

    Node* compositeHead = nullptr;
    Node* compositeTail = nullptr;

    Node* onesHead = nullptr;
    Node* onesTail = nullptr;

    Node* current = head;

    while (current) {
        Node* nextNode = current->next;
        current->next = nullptr;     

        if (current->value == 1) {
            if (!onesHead) onesHead = onesTail = current;
            else onesTail = onesTail->next = current;
        }
        else if (isPrime(current->value)) {
            if (!primeHead) primeHead = primeTail = current;
            else primeTail = primeTail->next = current;
        }
        else {
            if (!compositeHead) compositeHead = compositeTail = current;
            else compositeTail = compositeTail->next = current;
        }

        current = nextNode;
    }

    Node* newHead = nullptr;
    Node* mergeTail = nullptr;

    if (primeHead) {
        newHead = primeHead;
        mergeTail = primeTail;
    }

    if (compositeHead) {
        if (!newHead) newHead = compositeHead;
        if (mergeTail) mergeTail->next = compositeHead;
        mergeTail = compositeTail;
    }

    if (onesHead) {
        if (!newHead) newHead = onesHead;
        if (mergeTail) mergeTail->next = onesHead;
    }

    return newHead;
}

Node* buildList(const vector<int>& values) {
    Node* head = nullptr;
    Node* tail = nullptr;
    for (int v : values) {
        Node* newNode = new Node(v);
        if (!head) head = tail = newNode;
        else tail = tail->next = newNode;
    }
    return head;
}

void printList(Node* head) {
    while (head) {
        cout << head->value << (head->next ? " -> " : "");
        head = head->next;
    }
    cout << "\n";
}

int main() {
    int listSize;

    cout << "Enter number of nodes: ";
    cin >> listSize;

    vector<int> inputValues(listSize);
    cout << "Enter values:\n";
    for (int &x : inputValues) cin >> x;

    Node* head = buildList(inputValues);

    cout << "\nOriginal List:\n";
    printList(head);

    head = partitionPrimes(head);

    cout << "\nRearranged List (Primes → Composites → Ones):\n";
    printList(head);

    return 0;
}

File: week3Story/Story_3/Task5.java
/*
*Filename - Task5.java
*Description - You are given the head of a singly linked list of integers. Write a function
    partitionPrimes(head) that rearranges the list such that all nodes with prime values come
    before all nodes with composite values. The number 1 is considered neither prime nor
    composite and should be placed at the very end of the list.
    You must maintain the original relative order of the primes among themselves, and the
    composites among themselves. This must be done in-place (O(1) extra space).
Author - Nancy Jain
 */

import java.util.*;

public class Task5 {

    // ---------------- SAFE READ INT ----------------
    public static int readInt(Scanner sc) {
        while (true) {
            try {
                return Integer.parseInt(sc.next());
            } catch (Exception e) {
                System.out.print("Invalid input! Enter again: ");
            }
        }
    }

    // ---------------- LINKED LIST NODE -------------
    static class Node {
        int value;
        Node next;

        Node(int value) {
            this.value = value;
            this.next = null;
        }
    }

    // ---------------- CHECK PRIME -------------------
    static boolean isPrime(int number) {

        if (number <= 1) return false;
        if (number == 2) return true;
        if (number % 2 == 0) return false;

        for (int d = 3; d * d <= number; d += 2) {
            if (number % d == 0)
                return false;
        }
        return true;
    }

    // ----------- PARTITION INTO PRIME/COMPOSITE/ONES --------
    static Node partitionList(Node head) {

        Node primeHead = null, primeTail = null;
        Node compositeHead = null, compositeTail = null;
        Node onesHead = null, onesTail = null;

        Node current = head;

        while (current != null) {
            Node next = current.next;
            current.next = null; // detach

            if (current.value == 1) {
                // Ones list
                if (onesHead == null) onesHead = onesTail = current;
                else {
                    onesTail.next = current;
                    onesTail = current;
                }
            } 
            else if (isPrime(current.value)) {
                // Prime list
                if (primeHead == null) primeHead = primeTail = current;
                else {
                    primeTail.next = current;
                    primeTail = current;
                }
            } 
            else {
                // Composite list
                if (compositeHead == null) compositeHead = compositeTail = current;
                else {
                    compositeTail.next = current;
                    compositeTail = current;
                }
            }

            current = next;
        }

        // Combine lists: Primes → Composites → Ones
        Node finalHead = null, finalTail = null;

        if (primeHead != null) {
            finalHead = primeHead;
            finalTail = primeTail;
        }

        if (compositeHead != null) {
            if (finalHead == null) finalHead = compositeHead;
            else finalTail.next = compositeHead;
            finalTail = compositeTail;
        }

        if (onesHead != null) {
            if (finalHead == null) finalHead = onesHead;
            else finalTail.next = onesHead;
        }

        return finalHead;
    }

    // ---------------- BUILD LIST ----------------------
    static Node buildList(int[] arr) {
        Node head = null, tail = null;
        for (int num : arr) {
            Node newNode = new Node(num);
            if (head == null) head = tail = newNode;
            else {
                tail.next = newNode;
                tail = newNode;
            }
        }
        return head;
    }

    // ---------------- PRINT LIST ----------------------
    static void printList(Node head) {
        while (head != null) {
            System.out.print(head.value);
            if (head.next != null) System.out.print(" -> ");
            head = head.next;
        }
        System.out.println();
    }

    // ---------------- MAIN ----------------------------
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        while (true) {
            System.out.print("Enter number of nodes: ");
            int n = readInt(sc);

            int[] arr = new int[n];
            System.out.println("Enter values:");
            for (int i = 0; i < n; i++) {
                arr[i] = readInt(sc);
            }

            Node head = buildList(arr);

            System.out.println("\nOriginal List:");
            printList(head);

            head = partitionList(head);

            System.out.println("\nRearranged List (Primes → Composites → Ones):");
            printList(head);

            System.out.print("\nDo you want to continue? (yes-1 / no-0): ");
            int ch = readInt(sc);

            if (ch == 0) {
                System.out.println("Exiting…");
                break;
            }
        }
    }
}
